src/shared/utilities/timeoutUtils.ts

```diff
@@ -118,3 +149,62 @@ export async function waitUntil<T>(
         await new Promise(r => setTimeout(r, opt.interval))
     }
 }
+
+/** Best approximation for identifying a Promise in CommonJS */
+function isPromiseLike<T>(obj: any): obj is Promise<T> {
+    return obj !== undefined && (obj instanceof Promise || obj.then === 'function')
+}
+
+/**
+ * Race a Timeout object against a Promise. Handles Timeout expiration and cancellation, exposing access through
+ * the use of callbacks. Timeout tokens are cleaned up automatically after completion. Set `opt.completeTimeout`
+ * to false if this is not desired.
+ *
+ * @param promise Promise or a function that evaluates to a promise
+ * @param timeout Timeout token that will race against the promise
+ * @param opt.allowUndefined Output promise can resolve undefined (default: true)
+ * @param opt.onExpire Callback for when the promise expired. The callback can return a value
+ * @param opt.onCancel Callback for when the promise was cancelled. The callback can return a value
+ * @param opt.completeTimeout Automatically completes the Timeout upon resolution (default: true)
+ *
+ * @returns A Promise that returns if successful, or rejects when the Timeout was cancelled or expired.
+ */
+export function waitTimeout<T>(
+    promise: Promise<T> | (() => Promise<T>),
```

Comment:
should we rename this to `fn`
```suggestion
    fn: Promise<T> | (() => Promise<T>),
```

Reply:
Well the end result is a promise so I think promise makes the most sense.